### PR TITLE
Add support for multiple targets per remote secret.

### PR DESCRIPTION
**Pull request by metlos**
### What does this PR do?
This modifies the `RemoteSecret` to accept multiple targets.

Pros:
* the model is simple, doesn't require any new logic in persistence, etc.

Cons:
* as with the previous design, this does not allow for overriding values per target (but this is not a requirement at the moment).
* disallows multiple targets per namespace (this is not a big problem, because frankly, that is most probably just a corner case that is not likely to happen in practice).

This is just a draft PR because of the insufficient test coverage and lack of any manual testing.

### What issues does this PR fix or reference?
https://issues.redhat.com/browse/SVPI-477

### How to test this PR?
1. create a remotesecret with 2 targets, each one to a different namespace
2. check that the secret is created in each of the namespaces

```
quay.io/redhat-appstudio/pull-request-builds:spi-controller-593
quay.io/redhat-appstudio/pull-request-builds:spi-oauth-593

or

quay.io/lkrejci/service-provider-integration-operator:svpi-477
quay.io/lkrejci/service-provider-integration-oauth:svpi-477

```

### CODE REVIEW

**Comment by gazarenkov** [comment]
I like the API (it will be extended with cluster info in the future but fine for the time).
I'd prefer it to be merged better sooner than later (even if with low test coverage) to let other play with it.

File: controllers/remotesecret_controller.go [hunk 1]
```diff
@@ -156,11 +168,54 @@ func (r *RemoteSecretReconciler) Reconcile(ctx context.Context, req reconcile.Re
...
+			lg.Error(err, "failed to update the status of the remote secret with success after deploying everything successfully")
+			returnedError = err
+		}
+	}
+
+	return ctrl.Result{}, returnedError
+}
+
+func (r *RemoteSecretReconciler) processTargets(ctx context.Context, remoteSecret *api.RemoteSecret, secretData *remotesecretstorage.SecretData, errorAggregate *AggregatedError) {
+	namespaceClassification := remotesecrets.ClassifyTargetNamespaces(remoteSecret)
```
Comment (sparkoo) [inline controllers/remotesecret_controller.go:197]: How this indexing work? Is order guaranteed in k8s objects ?

**Review by sparkoo — COMMENTED** [review]
RemoteSecret in samples is not valid
```
λ k apply -f samples/remote-secret.yaml 
Error from server (BadRequest): error when creating "samples/remote-secret.yaml": RemoteSecret in version "v1beta1" cannot be handled as a RemoteSecret: strict decoding error: unknown field "spec.target"
```

Code is quite hard to read and follow, but it may be just my issue

File: api/v1beta1/remotesecret_types.go [hunk 2]
```diff
@@ -92,3 +107,17 @@ type RemoteSecretList struct {
 func init() {
 	SchemeBuilder.Register(&RemoteSecret{}, &RemoteSecretList{})
 }
+
+// Validate makes sure that no two targets specify the same namespace.
+func (rs *RemoteSecret) Validate() error {
+	nss := map[string]int{}
+
+	for i, t := range rs.Spec.Targets {
+		previous, present := nss[t.Namespace]
```
[the remark this answers is not in the document]
Comment (metlos) [reply api/v1beta1/remotesecret_types.go:116]: yep, fixed :)

File: controllers/remotesecret_controller.go [hunk 3]
```diff
@@ -156,11 +168,54 @@ func (r *RemoteSecretReconciler) Reconcile(ctx context.Context, req reconcile.Re
...
+		if err = r.saveSuccessStatus(ctx, remoteSecret); err != nil {
+			lg.Error(err, "failed to update the status of the remote secret with success after deploying everything successfully")
+			returnedError = err
+		}
+	}
+
+	return ctrl.Result{}, returnedError
+}
+
+func (r *RemoteSecretReconciler) processTargets(ctx context.Context, remoteSecret *api.RemoteSecret, secretData *remotesecretstorage.SecretData, errorAggregate *AggregatedError) {
```
[the remark this answers is not in the document]
Comment (metlos) [reply controllers/remotesecret_controller.go:196]: so that we can process as much as we can and aggregate what we couldn't.

File: controllers/remotesecret_controller.go (hunk 1 above)
Comment (metlos) [reply controllers/remotesecret_controller.go:197]: it is explained in the doc of the `ClassifyTargetNamespaces` function.

**Review by xbaran4 — COMMENTED** [review]
lgtm. will take a look again once complete
